src/Functions/FunctionsHashing.h

```diff
@@ -413,23 +414,47 @@ struct MurmurHash3Impl128
     static constexpr bool use_int_hash_for_pods = false;
 };
 
-/// http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/478a4add975b/src/share/classes/java/lang/String.java#l1452
 /// Care should be taken to do all calculation in unsigned integers (to avoid undefined behaviour on overflow)
 ///  but obtain the same result as it is done in signed integers with two's complement arithmetic.
 struct JavaHashImpl
 {
     static constexpr auto name = "javaHash";
     using ReturnType = Int32;
 
-    static Int32 apply(const char * data, const size_t size)
+    static ReturnType apply(int64_t x)
+    {
+        return static_cast<ReturnType>(
+            static_cast<uint32_t>(x) ^ static_cast<uint32_t>(static_cast<uint64_t>(x) >> 32)));
+    }
+
+    template <class T, typename std::enable_if<std::is_same_v<T, int8_t>
+                                                   || std::is_same_v<T, int16_t>
+                                                   || std::is_same_v<T, int32_t>, T>::type * = nullptr>
+    static ReturnType apply(T x)
+    {
+        return x;
+    }
+
+    template <typename T, typename std::enable_if<!std::is_same_v<T, int8_t>
+                                                      && !std::is_same_v<T, int16_t>
+                                                      && !std::is_same_v<T, int32_t>
+                                                      && !std::is_same_v<T, int64_t>, T>::type * = nullptr>
+    static ReturnType apply(T x)
+    {
+        const size_t size = sizeof(T);
+        const char * data = reinterpret_cast<const char *>(&x);
+        return apply(data, size);
```

Comment:
Also not sure about this code.

Now we apply javaHash for unsigned integers as for strings for bytes in the native byte order.

How unsigned integers are typically represented in Java?
If they are represented as BigInt, let's check what the behavior should be.
If they don't exist, let's throw an exception.

Reply:
Thanks for advice.  Java does not have unsigned types.